[routingconnector] convert routing processor to a connector

mwear opened the pull request:
**Description:** 

This PR introduces a connector version of the routing processor. It currently supports routing based on resource attributes OTTL statements only. Context based routing will be added later in a followup PR. 

There are two issues regarding fanout consumers that are being addressed in the core repo.
* The routing connector needs to be a consumer in multiple pipelines (the routing processor can route to a single exporter).
  * Will be resolved by: https://github.com/open-telemetry/opentelemetry-collector/pull/7840
* We need the ability to create connector routers to facilitate testing. I had to temporarily copy the fanoutconsumer package into the routing connector codebase due to package visibility issues.
  * Will be resolved by: https://github.com/open-telemetry/opentelemetry-collector/pull/7673 

We can address these issues as the relevant PRs land on the core repo.

cc: @djaglowski, @jpkrohling, @kovrus 

**Link to tracking Issue:** 
#19738

**Testing:** 
Unit / manual

**Documentation:**
The readme has been ported from the routing processor

mwear committed:
setup scaffolding for routingconnector

mwear committed:
add trace routing

mwear committed:
add config test

mwear committed:
add factory test

mwear committed:
add metric routing

mwear committed:
add log routing

mwear committed:
add more config tests

mwear committed:
use consistent conventions in test names

mwear committed:
formatting

mwear committed:
add readme

mwear committed:
lint

mwear committed:
update to use OTTL functions factory

mwear committed:
add changelog entry

mwear committed:
add codeowners

mwear committed:
update dep version

mwear committed:
fix imports

mwear committed:
add dependabot entry

djaglowski commented on connector/routingconnector/logs.go:
```diff
@@ -0,0 +1,145 @@
+
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		rlogs := ld.ResourceLogs().At(i)
+		ltx := ottllog.NewTransformContext(
+			plog.LogRecord{},
+			pcommon.InstrumentationScope{},
+			rlogs.Resource(),
+		)
```
Should we be using `ottlresource.NewTransformContext`?

mwear replied:
I made this change in c7e8a2bf74261f8731bda4d17eaf99602dbe4a37. Instead of using `ottllog.TransformContext`, `ottlspan.TransformContext`, etc we're now using `ottlresource.TransformContext` everywhere. This allowed me to remove the second type parameter from the router. It also means our statements can refer to `attributes` instead of `resource.attributes`. If you don't mind, take a look at the commit and let me know if that's what you had in mind. It simplifies the code significantly, so long as we're not losing anything. The choice to use the signal specific `TransformContext`s was inherited from the routing processor.

mwear replied:
cc: @jpkrohling @kovrus to double check this change

djaglowski replied:
Looks right to me. We'll end up needing the signal specific contexts for other conditions but for routing based on resources, I think we can and should keep it simple.

djaglowski commented on connector/routingconnector/metrics.go:
```diff
@@ -0,0 +1,148 @@
+			rmetrics.Resource(),
+		)
+
+		matchCount := len(c.router.routes)
```
What's the value of using an decrementing an int here, rather than just setting a bool when we see a match?

mwear replied:
Changed to a bool

djaglowski commented on connector/routingconnector/logs.go:
```diff
@@ -0,0 +1,145 @@
+	// the same set of exporters.
+	// This way we're not ending up with all the logs split up which would cause
+	// higher CPU usage.
+	groups := map[string]logsGroup{}
```
Do we need to keep track of groups according to a string key? Can we use the actual consumer? If so, I think this will simplify the `group` function.

mwear replied:
I updated to group by consumer. It allowed me to remove the spanGroup, logGroup, etc types altogether.

djaglowski reviewed (commented):
@mwear, thanks for taking this on. I'm excited to see so much progress.

mwear committed:
Apply suggestions from code review

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

mwear committed:
use []component.ID instead of []string for pipeline config

mwear committed:
group by consumer

mwear committed:
use bool instead of decrementing int

mwear committed:
use ottlresource.TransformContext instead of telemetry specific types

mwear committed:
update deps post rebase

mwear commented on connector/routingconnector/README.md:
```diff
@@ -0,0 +1,99 @@
+- Currently, it is not possible to specify boolean statements without function invocation as the routing condition. It is required to provide the NOOP `route()` or any other supported function as part of the routing statement, see [#13545](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13545) for more information.
+- Supported [OTTL] functions:
+  - [IsMatch](../../pkg/ottl/ottlfuncs/README.md#IsMatch)
+  - [delete_key](../../pkg/ottl/ottlfuncs/README.md#delete_key)
+  - [delete_matching_keys](../../pkg/ottl/ottlfuncs/README.md#delete_matching_keys)
```
It does. This works the same way that the routingprocessor does. Both the processor and the connector set `MutatesData: false`. I think the idea is that you can have an attribute used for routing only, and ensure that it does not get exported.

djaglowski replied:
👍 Makes sense

mwear committed:
lint

mwear commented:
@djaglowski thanks for the initial round of feedback. I was wondering if you had any opinions on the two issues I ran into with the fanoutconsumer (mentioned in the description). I could work around the package visibility issue by introducing  mock fanout consumers for the tests, but the actual fanout consumer is subtly complex, which introduces some risk. I have worked around the second issue, needing a connector be a receiver in multiple pipelines in order to get a fanout consumer, by adding that restriction to the routing connector; the processor did not have this restriction. Is that the right thing to do, or can we add the cabability to request a fanout consumer for a connector even if it is only a receiver in a single pipeline, the advantage being, that we can still request a consumer by name.

Let me know if you have opinions on those issues. I was waiting to get some resolution on them before taking this out of draft.

djaglowski commented:
@mwear, regarding problem 1 noted above:

I see what you are saying. I think what it boils down to is that we need a way for tests to create a `connector.TracesRouter` (and same for metrics/logs of course).

If we move `fanoutconsumer` from `service/internal` to `internal`, then `connectortest` could provide a minimal surface area to produce a "real" router that is only sufficient for testing. Something like the following - WDYT?

```go
type tracesRouterTestOption struct {
	id component.ID
	cons consumer.Traces
}

func WithNop(id component.ID) tracesRouterTestOption {
	return tracesRouterTestOption{ id: id, cons: consumertest.NewNop() }
}

func WithTracesSink(id component.ID) tracesRouterTestOption {
	return tracesRouterTestOption{ id: id, cons: &consumertest.TracesSink{} }	
}

func NewTracesRouterSink(opts ...tracesRouterTestOption) connector.TracesRouter {
	consumers := make(map[component.ID]consumer.Traces, numRoutes)
	for _, opt := range opts {
		consumers[opt.id] = opt.cons
	}
	return fanoutconsumer.NewTracesRouter(consumers)
}
```

djaglowski commented:
Regarding problem 2:

I tend to think that a router really should expect a minimum of two downstream pipelines. It seems that anything that could be done with a router that emits to only one pipeline, could (and should) be done in a simpler way.

mwear commented:
@djaglowski: Thanks for the suggestions. I like your proposal for problem 1 and will work on a PR it. I'm fine with leaving problem 2 as is. It makes sense to require multiple downstream pipelines, even though it introduce a slight difference in behavior for the connector vs the processor.

mwear commented:
I'll leave this in draft until there is some resolution on: https://github.com/open-telemetry/opentelemetry-collector/pull/7673. I'll also be OOO next week (back on May 22nd). I welcome any feedback in the meantime, and will address it when I return.

kovrus commented on connector/routingconnector/README.md:
```diff
@@ -0,0 +1,99 @@
+
+## Differences between the Routing Connector and Routing Processor
+
+- The connector will only route using [OTTL] statements which can only be applied to resource attributes. It does not support matching on context values at this time.
```
It seems that quite some users do routing based on the context values. Maybe after this PR we can follow up with the  OTTL function that extracts values from context if it is possible.

mwear replied:
We can add context based routing as a follow up, but we should discuss how we'd like that to work so that we don't paint ourselves into a corner with the first pass.

kovrus commented on connector/routingconnector/config.go:
```diff
@@ -0,0 +1,96 @@
+		}
+
+		if len(item.Pipelines) == 0 {
+			return fmt.Errorf("invalid route %s: %w", item.Value, errNoPipelines)
```
```suggestion
			return fmt.Errorf("invalid route %s: %w", item.Pipelines, errNoPipelines)
```

jpkrohling commented:
I'm way behind my github notifications, but I'm placing this to the top of my queue.

mwear committed:
Update connector/routingconnector/config.go

Co-authored-by: Ruslan Kovalov <ruslan.kovalov@grafana.com>

mwear committed:
remove unused `value` from config

djaglowski commented:
This PR is waiting for @bogdandrutu's response here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21498#issuecomment-1577924777

mwear commented:
> @mwear, I think we should move forward with the temporarily copied fanout consumer code. WDYT?

That works for me. Removing the copied code will be easy once the other changes land on core. Should I dust this off and bring it out of draft?

mwear committed:
update deps

mwear committed:
use shortform license

mwear committed:
lint

mwear committed:
generate issue templates

mwear committed:
use mdatagen to generate status

jpkrohling commented on connector/routingconnector/config.go:
```diff
@@ -0,0 +1,81 @@
+func (c *Config) Validate() error {
+	// validate that there's at least one item in the table
+	if len(c.Table) == 0 {
+		return fmt.Errorf("invalid routing table: %w", errNoTableItems)
```
Both parts of this error are static. Why not just return `errNoTableItems`?

mwear replied:
I restructured this error (and two others) in this method. Let me know if that's what you had in mind.

jpkrohling commented on connector/routingconnector/config.go:
```diff
@@ -0,0 +1,81 @@
+	// at least one pipeline
+	for _, item := range c.Table {
+		if len(item.Statement) == 0 {
+			return fmt.Errorf("invalid (empty) route : %w", errEmptyRoute)
```
Same to this and the next error.

jpkrohling commented on connector/routingconnector/internal/common/functions.go:
```diff
@@ -0,0 +1,28 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+)
+
+func createRouteFunction[K any](_ ottl.FunctionContext, _ ottl.Arguments) (ottl.ExprFunc[K], error) {
```
Probably a matter of taste, so, ignore me if you prefer how it is today.

```suggestion
func createRouteFunction[K any](ottl.FunctionContext, ottl.Arguments) (ottl.ExprFunc[K], error) {
```

jpkrohling commented on connector/routingconnector/internal/fanoutconsumer/logs.go:
```diff
@@ -0,0 +1,120 @@
+
+// Package fanoutconsumer contains implementations of Traces/Metrics/Logs consumers
+// that fan out the data to multiple other consumers.
+package fanoutconsumer // import "github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector/internal/fanoutconsumer"
```
Is this a copy of what we have at the fanoutconsumer in the core? If so, explaining why this was copied and what was changed would be good.

jpkrohling replied:
I just saw the mentions to this in the PR description. I think it needs to be in the code as well. Like: 

```
TODO: once we have fanoutconsumer test helpers, use that instead of copying the code
```

jpkrohling commented on connector/routingconnector/logs.go:
```diff
@@ -0,0 +1,116 @@
+
+		noRoutesMatch := true
+		for _, route := range c.router.routes {
+			_, isMatch, err := route.statement.Execute(ctx, rtx)
```
Is the error related to the data, or would it err only when the OTTL statement is wrong?

mwear replied:
The `Execute` method is part of the OTTL statement. Looking at [the source](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/4e082f40991a2353d2994b376fbda1a95d51a96e/pkg/ottl/parser.go#L49-L66), it will error if there is an error evaluating the condition, or the function in the statement. I _think_ this would indicate a problem with the statement.

jpkrohling replied:
Does this mean that this will always happen for all data points? Is it a case for a panic?

jpkrohling commented on connector/routingconnector/logs.go:
```diff
@@ -0,0 +1,116 @@
+
+		if noRoutesMatch {
+			// no route conditions are matched, add resource logs to default exporters group
+			c.group(groups, c.router.defaultConsumer, rlogs)
```
This might become clear later on, but isn't there a case where the only code branch reached was the error one, which made the defaultConsumer to be added already?

mwear replied:
The error conditions would come in evaluating the routes. If no routes match (or all of them error) then the logs should be sent to the default consumer.

jpkrohling commented on connector/routingconnector/logs.go:
```diff
@@ -0,0 +1,116 @@
+			c.group(groups, c.router.defaultConsumer, rlogs)
+		}
+	}
+	for consumer, group := range groups {
```
This confused me at first, as I didn't see groups being set, but then I realized that perhaps c.group is changing groups? If that's the case, wouldn't it be better to have c.group to return groups and set groups to it? It shouldn't cause new allocations and would make the operation more explicit.

mwear replied:
groups is a `map[consumer.Logs]plog.Logs`. it is mutated by `c.group`. since it is mutated by `c.group` it would be slightly confusing to me to also return and assign it. if we made that change, i would be under the impression that c.group returns a new map (without delving into the details of the method), which is not the case.

jpkrohling replied:
Makes sense, thanks!

jpkrohling commented on connector/routingconnector/logs_test.go:
```diff
@@ -0,0 +1,329 @@
+	v, ok := attrs.Get("attr")
+	assert.True(t, ok, "non routing attributes shouldn't be dropped")
+	assert.Equal(t, "acme", v.Str())
+	require.Len(t, sink0.AllLogs(), 0,
```
Shouldn't this be an assert instead? I would even argue that all requires in this part should be asserts, as this is in the verification phase of the test, but I see why the others are `require` (just not this one)

jpkrohling reviewed (approved):
LGTM, a couple of minor comments, nothing blocking. This is REALLY cool, thank you @mwear!

mwear commented on .github/dependabot.yml:
```diff
@@ -1097,8 +1102,3 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
-  - package-ecosystem: "gomod"
```
This was the result of running `make gendependabot`. I can look into it more.

mwear commented on connector/routingconnector/README.md:
```diff
@@ -0,0 +1,106 @@
+- The connector routes to pipelines, not exporters as the processor does.
+
+### OTTL Limitations
+- Currently, it is not possible to specify boolean statements without function invocation as the routing condition. It is required to provide the NOOP `route()` or any other supported function as part of the routing statement, see [#13545](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13545) for more information.
```
This came from the routing processor readme: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/routingprocessor#limitations.

mwear committed:
update deps

mwear committed:
restructure errors

mwear committed:
use assert

mwear committed:
add todo to use fanout consumer helpers when available

mwear committed:
remove tooFewConsumers error

Now that https://github.com/open-telemetry/opentelemetry-collector/pull/7840
merged, we can expect a fanout consumer to be passed consistently to a
connector factory. We still check for an error casting the consumer
to a connector router, but this would only happen if a breaking change
was introduced to the connector internals.

mwear commented on connector/routingconnector/README.md:
```diff
@@ -0,0 +1,106 @@
+## Differences between the Routing Connector and Routing Processor
+
+- The connector will only route using [OTTL] statements which can only be applied to resource attributes. It does not support matching on context values at this time.
+- The connector must be a receiver in a minimum of two pipelines.
```
This actually a limitation of connectors before https://github.com/open-telemetry/opentelemetry-collector/pull/7840 merged. Since that has been merged, and released, we no longer have that limitation. I've updated the PR.

mwear committed:
make gendependabot

because CI told me to:
> dependabot.yml is out of date, please run "make gendependabot" and commit the changes in this PR

mwear committed:
update deps

mwear committed:
remove fanoutconsumer; use connectortest helpers

mwear committed:
add all connector options to component validation

djaglowski commented:
A few comments have not been marked as resolved, but I believe all concerns have been addressed. @jpkrohling, @kovrus, please double check the open comments and raise issues if you would like to see additional changes.

@mwear, thank you for all the work on this! This is a critical connector and I'm looking forward to making it available to users.
